File: packages/urdu-js/src/transliteration/from-roman.ts
import { ROMAN_TO_URDU } from './roman-map.js'

export function fromRoman(text: string): string {
  if (!text) return text
  let result = '', i = 0
  while (i < text.length) {
    let matched = false
    for (const [roman, urdu] of ROMAN_TO_URDU) {
      if (text.startsWith(roman, i)) {
        result += urdu; i += roman.length; matched = true; break
      }
    }
    if (!matched) { result += text[i]!; i++ }
  }
  return result
}

File: packages/urdu-dart/lib/src/transliteration/from_roman.dart
import 'roman_map.dart';

/// Transliterates Roman Urdu text back to Urdu script based on a mapping table.
String fromRoman(String text) {
  if (text.isEmpty) return text;
  final buf = StringBuffer();
  var i = 0;
  while (i < text.length) {
    var matched = false;
    for (final entry in romanToUrdu) {
      final roman = entry.$1;
      final urdu = entry.$2;
      if (text.startsWith(roman, i)) {
        buf.write(urdu);
        i += roman.length;
        matched = true;
        break;
      }
    }
    if (!matched) {
      buf.write(text[i]);
      i++;
    }
  }
  return buf.toString();
}
